File: meeting-rooms-ii/meeting-rooms-ii.cpp
class Solution {
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {
        int rooms = 0;
        sort(intervals.begin(), intervals.end());
        priority_queue<int, vector<int>, greater<int>> pq;
        for(auto x: intervals){
            int start = x[0], end = x[1];
            while(!pq.empty() and pq.top() <= start){
                pq.pop();
            }
            pq.push(end);
            rooms = max(rooms, (int)pq.size());
        }
        return rooms;
    }
};

File: meeting-rooms-ii/meeting-rooms-ii.java
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        if(intervals.length == 0){
            return 0;
        }
        Arrays.sort(intervals, Comparator.comparing(a -> a[0]));
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        pq.add(intervals[0][1]);
        for(int i = 1; i< intervals.length; i++){
            if(intervals[i][0] >= pq.peek()){
                pq.remove();
            }
            pq.add(intervals[i][1]);
        }
        return pq.size();
    }
}